include/rx/ranges.hpp

```diff
@@ -2317,6 +2317,109 @@ template <class... Inputs>
 }
 
+/// Forward all elements of the range as-is.
+struct passthrough {
+    template <class R>
+    struct Range {
+        using output_type = get_output_type_of_t<R>;
+        static constexpr bool is_finite = is_finite_v<R>;
+        static constexpr bool is_idempotent = is_idempotent_v<R>;
+
+        R input;
+
+        constexpr void next() {
+            RX_ASSERT(!at_end());
+            input.next();
+        }
+
+        [[nodiscard]] constexpr decltype(auto) get() const {
+            RX_ASSERT(!at_end());
+            return input.get();
+        }
+
+        [[nodiscard]] constexpr bool at_end() const {
+            return input.at_end();
+        }
+
+        [[nodiscard]] constexpr size_t size_hint() const {
+            return input.size_hint();
+        }
+    };
+
+    template <class InputRange>
+    [[nodiscard]] constexpr auto operator()(InputRange&& input) const {
+        using Inner = get_range_type_t<InputRange>;
+        return Range<Inner>{as_input_range(std::forward<InputRange>(input))};
+    }
+};
+
+/// Present elements in a range to a function, then pass the values along as-is.
+template <class F>
+struct spy {
+    F func;
+
+    template <class Fx>
+    explicit constexpr spy(Fx&& func) : func(std::forward<Fx>(func)) {}
+
+    template <class R>
+    struct Range {
+        using output_type = get_output_type_of_t<R>;
+        static constexpr bool is_finite = is_finite_v<R>;
+        static constexpr bool is_idempotent = false; // everytime get() is called func() is called
+
+        R input;
+        spy s;
+
+        constexpr void next() {
+            RX_ASSERT(!at_end());
+            input.next();
+        }
+
+        [[nodiscard]] constexpr decltype(auto) get() const {
+            RX_ASSERT(!at_end());
+            auto&& value = input.get();
+            const auto& const_value = value;
+            s.func(const_value);
+            return std::forward<decltype(value)>(value);
+        }
+
+        [[nodiscard]] constexpr bool at_end() const {
+            return input.at_end();
+        }
+
+        [[nodiscard]] constexpr size_t size_hint() const {
+            return input.size_hint();
+        }
+    };
+
+    template <class InputRange>
+    [[nodiscard]] constexpr auto operator()(InputRange&& input) const& {
+        using Inner = get_range_type_t<InputRange>;
+        return Range<Inner>{as_input_range(std::forward<InputRange>(input)), *this};
+    }
+
+    template <class InputRange>
+    [[nodiscard]] constexpr auto operator()(InputRange&& input) && {
+        using Inner = get_range_type_t<InputRange>;
+        return Range<Inner>{as_input_range(std::forward<InputRange>(input)), std::move(*this)};
+    }
+};
+template <class F>
+spy(F&&)->spy<remove_cvref_t<F>>;
+
+/// A sink that simply discards all elements of a range.
+///
+/// Use as `append(void_sink())` if you need the range to run, but you don't need the result.
+struct void_sink {
```

Comment:
This seems useful. :+1: How about calling it `null_sink` instead? (It's like "/dev/null")